### PR TITLE
Per project synthesized assets

### PR DESCRIPTION
**Pull request by ryantstone**
Resolves https://github.com/tuist/tuist/issues/2249
Request for comments document (if applies):

### Short description 📝
Addresses #2249. Adds a `ProjectOptions` enum with currently a single case for enabling/disabling per-project synthesized resources. 

I've validated through testing and unit tests that it works as expected. I've had difficulty running acceptance tests due to xcodebuild errors which aren't clear enough as to if the test is failing due to a code change or because of some other issue.

### Checklist ✅
- [x] The code architecture and patterns are consistent with the rest of the codebase.
- [ ] The changes have been tested following the [documented guidelines](https://tuist.io/docs/contribution/testing-strategy/).
- [x] The `CHANGELOG.md` has been updated to reflect the changes. In case of a breaking change, it's been flagged as such.
- [ ] In case the PR introduces changes that affect users, the documentation has been updated.

### CODE REVIEW

**Review by kwridan — COMMENTED** [review]
Thanks for submitting this @ryantstone 

There are a few interesting behaviours we need to be mindful of, some of which could introduce breaking changes 

- As things stand `Config.swift` controls this particular mapper in [`ProjectMapperProvider`](
https://github.com/tuist/tuist/blob/main/Sources/TuistKit/GraphMappers/ProjectMapperProvider.swift#L38) - users may already have this set
- The config is opt out rather than opt in 
- The newly introduced option in `ProjectOptions` is opt in
- This will cause a conflict between both!

A suggestion to help make a smooth transition for existing users as well as enable the new functionality, possibly something like this could work:

- Maintain the existing opt out behaviour - so it's enabled by default, but the option can disable it (it would need to be renamed to make it clearer or perhaps follow @fortmarek's [suggestion](https://github.com/tuist/tuist/issues/2249#issuecomment-777042295) to contain the enabled state as an associated type
- Add a new mapper `DisableResourceAccessorsProjectMapper` which would add the `synthesizedResourceAccessors(false)` `ProjectOption` to all projects incase `Config.swift` has the `disableSynthesizedResourceAccessors` option set

>  I've had difficulty running acceptance tests due to xcodebuild errors which aren't clear enough as to if the test is failing due to a code change or because of some other issue.

Ouch, would be good to know what errors you're seeing. Sometimes it may be easier to manually test the change when iterating locally rather than running the entire acceptance test suite - to do so:

```sh
swift build
swift run tuist generate --path fixtures/ios_app_with_framework_and_resources`
```

then inspecting the generated workspace.

Hope this helps

**Comment by ryantstone** [comment]
@kwridan  Thanks for the help. I got the acceptance test figured out, I was passing an incorrect flag on `swift build` which caused it to build incorrectly. Which then caused the test to fail when running the acceptance tests. 

There was a follow-on discussion at https://github.com/tuist/tuist/discussions/2564#discussioncomment-431529 about how best to handle the API. Would having a default argument to enable synthesized resources be the right path? There aren't any current default argument parameters for `Project` so this would be breaking with the existing API of making no assumptions about what a developer intends, that said it would be reasonable to have it as automatic.

**Comment by kwridan** [comment]
> Would having a default argument to enable synthesized resources be the right path? There aren't any current default argument parameters for Project so this would be breaking with the existing API of making no assumptions about what a developer intends, that said it would be reasonable to have it as automatic.

That's a good point, I was thinking perhaps internally it could default to being enabled when the option is not specified to maintain the existing behaviour and the option can be used to opt out or explicitly opt in (incase the defaults changes). I could see the argument for either approach. 

cc @fortmarek & @luispadron - based on the current plans for the plugins support, are there any items that can be done at this stage to ease it / ensure compatibility?

**Review by fortmarek — COMMENTED** [review]
Hey @ryantstone! Thanks for this PR 🖤 Let me chime in:

We should probably take the plugins into consideration as well. As things stand, there is no way to specify for which type of resources you want which templates.

Consider the following scenario:
I have PluginA from which I want to use strings templates
I have PluginB from which I want to use custom lottie templates

But PluginB has also strings templates - how do we determine if we use tuist templates, PluginA templates, or PluginB templates?

I can think of two ways how to allow this (using what's been proposed [here](https://github.com/tuist/tuist/discussions/2604)):
Add `resourceSynthesizerTemplates` property to `Project` where you could specify something like:
```swift
resourceSynthesizerTemplates: [
    .strings(plugin: "PluginA"),
    .plists(.default),
    .custom(plugin: "PluginB", name: "Lottie") // name alludes to the name of the custom resource synthesizer in the plugin
]
```

But this would lead to yet-another-property which is why we wanted to go with the direction of options.

So, what we could do is change `synthesizedResourceAccessors` case to `synthesizedResourceAccessors(Set<ResourceSynthesizerTemplate>)` where the API would be similar to above:

```swift
options: [
   .synthesizedResourceAccessors([.strings(plugin: "PluginA")....])
]
```

you should also be able to opt out:
```swift
options: [
    .synthesizedResourceAccessors(.none) // or instead of .none .disable maybe? Or something else entirely?
]
```

To explicitly state you want all resource accessors, you could do `synthesizedResourceAccessors(.all)` or to use default `synthesizedResourceAccessors(.default)` (I'd probably just use one of these options as they'd currently mean the same thing).

Now, as far as this PR is concerned, we don't have anything like `ResourceSynthesizerTemplate` and I think it's out of scope. What we could do is to make a temporary type `ResourceSynthesizerFlag` (there could probably be a better name) and add it as associated value to `synthesizedResourceAccessors`, so you could write `.synthesizedResourceAccessors(.none)` (so it's possible to extend this with more fine-grained control). The issue with this is that users could quite possibly use the type itself in the helpers (as it happened with resources). That's why I think that we should fallback to `.disableSynthesizedResourceAccessors` which can later be deprecated and automatically converted to `.synthesizedResourceAccessors(.none)`, so we can ensure the change won't be breaking.

The issue with adding this as `options` is the following scenario:
```swift
options: [
   .synthesizedResourceAccessors(.none),
   .synthesizedResourceAccessors([.strings(plugin: "PluginA")]),
]
```

What should happen in this case? These two options invalidate each other, after all. 

We could probably address this in the later part of the process and throw an error if we have two resource accessors options that either contradict each other or it's ambiguous which one to use (eg if you specify `strings(plugin: "PluginA")` and `strings(plugin: "PluginB")`). I am not sure if there is a better way to do this in a static manner.

To sum up:
For the purpose of this PR, I am inclined to use `.disableSynthesizedResourceAccessors` which should make it fairly easy for us to change it to `synthesizedResourceAccessors(.none)` (or something similar) when plugins are available for resource accessors.

Let me know if this makes sense to you @ryantstone and @kwridan - and sorry for extending this with the somewhat lengthy plugins discussion which might seem a little off-topic here but we should take it into a consideration. 

cc @luispadron @fila95 this is related to resource accessors plugins

**Comment by ryantstone** [comment]
Would this be more appropriate to roll into the work @luispadron is doing or defer it until it is done so I have a better idea of what that API looks like? Looking through the information on plugins I'm not sure I could develop something in vacuum w/o having an existing API to code against.

**Comment by fortmarek** [comment]
> Would this be more appropriate to roll into the work @luispadron is doing or defer it until it is done so I have a better idea of what that API looks like?

I believe this might be the right call at the moment. I'll try to start working on this, so this can be resolved sooner rather than later. Sorry for the inconvenience.
